Upwork Support: Update plan checks

obenland opened the pull request:
I'm having a bit of trouble with this one.


#### Changes proposed in this Pull Request

* Updates plan checks with Pro plan.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

I'm not sure how to functionally test this because Happychat is only available in Production and the fallback is email support.

Hacky testing:
* Add a `console.log(supportVariation);` [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/blocks/inline-help/inline-help-contact-view.jsx#L16).
* Apply D80779-code to your sandbox and sandbox public-api.
* Create a new user and site and give it a personal or premium plan.
* Select a mag16 language other than English, like German
* Click on the help icon in the bottom right of the screen.
* Check console for `SUPPORT_UPWORK_TICKET` as the support variation.

Related to p4TIVU-a66-p2

obenland committed:
Upwork Support: First pass at feature check

obenland commented:
@mattwondra Is it expected that customers with Premium plans that contain live support would instead be re-routed to upwork support in cases where they don't also have a site with business plan and up?

I'm trying to make sense of this selector and how a feature check could work here but I'm not sure how.
I would have assumed I could just check if they have any site with live-support, but the Premium plan doesn't allow for that without breaking back compat.

mattwondra commented:
@obenland 👋 I'm no longer on the HappOps team, so @Automattic/happops-devs will be better suited to give current feedback on questions like this. Here's what I can tell you though:

- Upwork is a service that helps us support users in some non-English locales, by letting them get ticket (email) support in their own language through contracted translators.
- Users are eligible for Upwork if they are in one of our enabled locales.
- However, we don't use Upwork for users in our higher-paid plans, because we prefer to give them live chat support directly from our HEs. So users with Business/eCommerce plans are not eligible for Upwork ticket support.
- We offer users support based on their highest plan across all their sites. So you're correct, that if a user has a Premium plan, but also has Business/eCommerce plan, they are not eligible for Upwork (because we want them routed to live chat, even if they're asking a question about their Premium site).

Hope that's helpful but let me know if I can clarify anything else!

obenland commented:
Ah, apologies, thanks for providing that context regardless!

With the Premium plan offering live-chat support, should this check be extended to include that plan? If so we could check if any site has live-chat support to make the decision feature-based.

obenland committed:
Use new upwork exempt feature

We can't check for live-chatt support because Premium plans still get upwork support

mreishus commented on client/state/selectors/is-eligible-for-upwork-support.ts:
```diff
@@ -15,11 +16,7 @@ export default function isEligibleForUpworkSupport( state: AppState ): boolean {
-	return ! hasBusinessOrEcommercePlan;
+	return ! Object.values( getSitesItems( state ) ).some( ( { ID } ) =>
+		siteHasFeature( state, ID ?? 0, WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT )
+	);
```
It took me a moment to understand this logic, may I recommend a comment?

```diff
diff --git a/client/state/selectors/is-eligible-for-upwork-support.ts b/client/state/selectors/is-eligible-for-upwork-support.ts
index 82e3e78a59..e9a79d6741 100644
--- a/client/state/selectors/is-eligible-for-upwork-support.ts
+++ b/client/state/selectors/is-eligible-for-upwork-support.ts
@@ -16,6 +16,8 @@ export default function isEligibleForUpworkSupport( state: AppState ): boolean {
                return false;
        }
 
+       // If any site has the WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT feature, the
+       // user is not eligible for Upwork Support.
        return ! Object.values( getSitesItems( state ) ).some( ( { ID } ) =>
                siteHasFeature( state, ID ?? 0, WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT )
        );
```

or

```diff
diff --git a/client/state/selectors/is-eligible-for-upwork-support.ts b/client/state/selectors/is-eligible-for-upwork-support.ts
index 82e3e78a59..f0ab6f3bcb 100644
--- a/client/state/selectors/is-eligible-for-upwork-support.ts
+++ b/client/state/selectors/is-eligible-for-upwork-support.ts
@@ -16,7 +16,8 @@ export default function isEligibleForUpworkSupport( state: AppState ): boolean {
                return false;
        }
 
-       return ! Object.values( getSitesItems( state ) ).some( ( { ID } ) =>
-               siteHasFeature( state, ID ?? 0, WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT )
+       // To be eligible for Upwork, all of a user's sites must not have the WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT feature.
+       return Object.values( getSitesItems( state ) ).every(
+               ( { ID } ) => ! siteHasFeature( state, ID ?? 0, WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT )
        );
 }
```

klimeryk replied:
I'd prefer to avoid comments like this - they quickly go out of date and cause more confusion than help.

Would it be possible to rewrite the code so that it's more readable instead? Maybe just extracting the sites items to a variable would help?

```suggestion
const sites = Object.values( getSitesItems( state ) );
return ! sites.some( ( { ID } ) =>
		siteHasFeature( state, ID ?? 0, WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT )
	);
```

If that's not enough, we can extract the callback to a separate function (in the same file should be fine) so that the `return` statement would simply read:
```
return ! sites.some( hasUpworkSupportExempt );
```

obenland committed:
Add clarifying comment

Props @mreishus

mreishus commented:
Since Calypso by default only loads the selected site's features, on my main account with a variety of sites, I'm seeing different results based on what sites I've let calypso load.

Examples:

- `http://calypso.localhost:3000/stats/ads/day/<prosite>` --> "SUPPORT_CHAT_OVERFLOW" (Good)
- `http://calypso.localhost:3000/me` --> "SUPPORT_UPWORK_TICKET"  (Expected to see chat overflow)
- `http://calypso.localhost:3000/stats/ads/day/<freesite>` ---> "SUPPORT_UPWORK_TICKET" (Expected to see chat overflow)
- `http://calypso.localhost:3000/stats/ads/day/<freesite>` -> Switch sites to a Pro site -> Switch sites back to a free site ----> "SUPPORT_CHAT_OVERFLOW" (Good)

obenland commented:
whoa, nice catch! It looks like that's based on `getSitesItems()` returning an empty object, so this might not be a regression?
Should components that use the selector query all sites when they mount, to fix that?

mreishus reviewed (approved):
Ah, right, it's not a regression since the same bug existed before. The logic in this PR is working well in combination with D80779-code (which should be deployed before this PR is).

> Should components that use the selector query all sites when they mount, to fix that?

I think it could be a good fix. One danger would be we'd want to make sure we don't have the opposite problem (fetching features too often). For `<InlineHelpContactView>` I don't think that's a danger, but I'm not familiar with the other uses of `getSitesItems`.

klimeryk commented on client/state/selectors/test/is-eligible-for-upwork-support.js:
```diff
@@ -13,46 +7,51 @@ describe( 'isEligibleForUpworkSupport()', () => {
+						222: { ID: 222 },
+					},
+					features: {
+						111: { data: { active: [ 'upwork-support-exempt' ] } },
```
Can we use the `WPCOM_FEATURES_UPWORK_SUPPORT_EXEMPT` constant here instead? That will ensure no magic string and it will make the code more future-proof (if we want to change the underlying string or the feature altogether).

klimeryk commented on client/state/selectors/test/is-eligible-for-upwork-support.js:
```diff
@@ -13,46 +7,51 @@ describe( 'isEligibleForUpworkSupport()', () => {
+					},
+					features: {
+						111: { data: { active: [ 'upwork-support-exempt' ] } },
+						222: { data: { active: [ 'live-support' ] } },
```
Similar comment as above about using a constant - and it actually seems to show the issue. The actual string should be `live_support` 😅 (yay for no consistency - of course, not your fault, these have been established long time ago)
https://github.com/Automattic/wp-calypso/blob/09c24573994ff390e63f6af24e679a73c002925c/packages/calypso-products/src/constants/features.ts#L222

klimeryk commented on client/state/selectors/test/is-eligible-for-upwork-support.js:
```diff
@@ -13,46 +7,51 @@ describe( 'isEligibleForUpworkSupport()', () => {
+						222: { ID: 222 },
+					},
+					features: {
+						111: { data: { active: [ 'live-chat' ] } },
```
Not sure we have a `live-chat` feature at all - did you mean `live_support` or one of the `live-chat-*` ones?

klimeryk reviewed (changes requested):
Thanks for improving the check - glad to see more feature checks (versus plan, upgrades, etc. checks that cause more work long-term).

However, as @mreishus mentioned in https://github.com/Automattic/wp-calypso/pull/63623#issuecomment-1128058725, there's indeed a regression here.

To reproduce, I checked a test user of mine with Spanish locale. On production, they get live chat (or live chat overflow) support variations. However, on this PR, they get the Upwork variation.
That's because we check now for the features flags - which are loaded from a separate call _per site_. The previous check was using the `plan` object that is returned for the more general `sites` call. In other words, we always have plan info for all sites, but we only have features for a given site if we explicitly loaded it first. So the check "if any site has this feature" is pretty flawed, cause often it will only be operating on very partial data.

As a result, this causes a regression and crucial change of behavior. This is a blocker for deploying this as-is. I'm not sure what's the fix here. Adding features to the sites endpoint? It's already a hot spot and I'd be wary of adding more stuff there 🤔 But if we already have plan info there, maybe it's possible.That's definitely a scope creep, though 😓

mreishus commented:
Something like this queries features for all sites:
```jsx
import { useSelector } from 'react-redux';
import QuerySiteFeatures from 'calypso/components/data/query-site-features';
import getSites from 'calypso/state/selectors/get-sites';

export default function QueryAllSiteFeatures() {
	const sites = useSelector( getSites );
	const allSiteIds = sites.map( ( site ) => site.ID );
	return <QuerySiteFeatures siteIds={ allSiteIds } />;
}

// <QueryAllSiteFeatures />
```
Unfortunately, it kicks off one HTTP request per site, which makes for 100+ requests for me, since I have so many sites:
![2022-05-18_14-30](https://user-images.githubusercontent.com/937354/169141062-0d09fb6a-9f32-4427-b83c-ee51d5d98557.png)

**Ideas**:
* Modify sites endpoint to return features
* Modify features endpoint to allow multiple sites to be queried in one request; strategically use `<QueryAllSiteFeatures>` in files like `client/me/help/help-contact/index.jsx`†
* Modify features endpoint to add a "return all features for all my sites" function without having to specify siteids; strategically use `<QueryAllSiteFeatures>`
* Make a new API endpoint that calculates the support level and returns that value without having to send all the features to calypso
* Other?

----

†_I see these files as needing access to the 'features for all sites' data:_

- `client/me/help/help-contact/index.jsx`
- `client/blocks/inline-help/inline-help-contact-view.jsx`
- `client/components/marketing-survey/cancel-purchase-form/index.jsx`
- `client/my-sites/checkout/composite-checkout/components/checkout-help-link.tsx`

obenland commented:
> Make a new API endpoint that calculates the support level and returns that value without having to send all the features to calypso

I like this one the best. With support levels being user-based they'll always be tough to map to site features

obenland committed:
Revert to using plan checks

obenland commented:
@klimeryk I decided to forego any more foundational changes here and accept that support will need its own (new) approach

klimeryk commented on client/state/selectors/is-eligible-for-upwork-support.ts:
```diff
@@ -15,11 +15,9 @@ export default function isEligibleForUpworkSupport( state: AppState ): boolean {
-	const hasBusinessOrEcommercePlan = Object.values( getSitesItems( state ) ).some( ( site ) => {
-		const planSlug = site.plan?.product_slug;
-		return isBusinessPlan( planSlug ?? '' ) || isEcommercePlan( planSlug ?? '' );
+	// If any site has an Upwork Support exemption, the user is not eligible.
```
See what I mean be comments getting obsolete? ;) Even before deploy 😬 
(I'm half-joking - this patch has been through so many iterations, that I fully understand a leftover of previous version got through)
I'd vote for simply no comment - the code seems straightforward enough.

klimeryk reviewed (approved):
Thanks for bearing with this code and some legacy decisions 🙇 Apologies for the possibly frustrating experience here.
I've tested and works as expected 👍

obenland committed:
Remove unnecessary comment.

obenland commented:
> Apologies for the possibly frustrating experience here.

Not at all! Thanks for taking the time and preventing me from heading down the wrong path!
